lib/roar/json/json_api/declarative.rb

```diff
@@ -0,0 +1,100 @@
+module Roar
+  module JSON
+    module JSONAPI
+      # New API for JSON-API representers.
+      module Declarative
+        def type(name = nil)
+          return @type unless name # original name.
+
+          heritage.record(:type, name)
+          @type = name.to_s
+        end
+
+        def link(name, options = {}, &block)
+          return super(name, &block) unless options[:toplevel]
+          for_collection.link(name, &block)
+        end
+
+        def meta(options = {}, &block)
+          return super(&block) unless options[:toplevel]
+          for_collection.meta(&block)
+        end
+
+        def relationship(options = {}, &block); end
+
+        def has_one(name, options = {}, &block)
+          has_relationship(name, options.merge(collection: false), &block)
+        end
+
+        def has_many(name, options = {}, &block)
+          has_relationship(name, options.merge(collection: true), &block)
+        end
+
+        private
+
+        def has_relationship(name, options = {}, &block)
+          # every nested representer is a full-blown JSONAPI representer.
+          nested(:included, inherit: true) do
+            property(name, collection: options[:collection]) {
+              include Roar::JSON::JSONAPI
+
+              instance_exec(&block)
+
+              def from_document(hash)
+                hash
+              end
+            }
+          end
+
+          resource_identifier_representer = Class.new(Roar::Decorator)
+          resource_identifier_representer.class_eval do
+            include Roar::JSON
+            include Roar::Hypermedia
+            include JSONAPI::Meta
+            extend JSONAPI::Declarative
+
+            def self.relationship_block
+              @relationship_block || -> {}
```

Comment:
(`relationship` is conceptually similar, but not exactly the same as the current `toplevel` option for links, etc. - hence different approach)